File: Results/DER/Translation/translation_output/w_test/granite/codenet/python/java-sanitized/p03088.java
import java.util.*;

public class p03088 {
    static int N, MOD = 100000007;
    static Map<String, Integer>[][] memo;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        memo = new HashMap[N + 1][4];

        System.out.println(dfs(0, "TTT"));
        sc.close();
    }

    public static boolean ok(String last4) {
        for (int i = 0; i < 4; i++) {
            // tã®åæå
            char[] t = last4.toCharArray();

            if (i >= 1) {
                // æå­ã®å¥ãæ¿ã
                char temp = t[i - 1];
                t[i - 1] = t[i];
                t[i] = temp;
            }

            if (new String(t).contains("AGC")) {
                return false;
            }
        }

        return true;
    }

    public static int dfs(int cur, String last3) {
        // æ·±ãã¨ããããæ»ã£ã¦ããéç¨ã§
        // ä»ã¾ã§ã®last3 ã®ãã¿ã¼ã³ã¨åæ§ãªãåå©ç¨
        if (memo[cur].containsKey(last3)) {
            return memo[cur].get(last3);
        }

        if (cur == N) {
            return 1;
        }
        int ret = 0;

        for (char c : "ACGT".toCharArray()) {
            if (ok(last3 + c)) {
                // ok ã®å ´åcount ãå¢ãã
                // 1åãããã¦ curãå¢ãã
                ret = (ret + dfs(cur + 1, last3.substring(1) + c)) % MOD;
            }
        }
        memo[cur].put(last3, ret);

        return ret;
    }
}

File: dataset/Intermediate/Translation/codenet/python/code/p03088.py
N, MOD = int(input()), 10 ** 9 + 7
memo = [{} for i in range(N + 1)]


def ok(last4):
    for i in range(4):
        # tの初期化
        t = list(last4)

        if i >= 1:
            # 文字の入れ替え
            t[i - 1], t[i] = t[i], t[i - 1]

        if "".join(t).count("AGC") >= 1:
            return False

    return True


def dfs(cur, last3):
    # 深いところから戻っていく過程で
    # 今までのlast3 のパターンと同様なら再利用
    if last3 in memo[cur]:
        return memo[cur][last3]

    if cur == N:
        return 1
    ret = 0

    for c in "ACGT":
        if ok(last3 + c):
            # ok の場合count を増やす
            # 1個ずらして curを増やす
            ret = (ret + dfs(cur + 1, last3[1:] + c)) % MOD
    memo[cur][last3] = ret

    return ret


print(dfs(0, "TTT"))
